src/diffusers/models/vq_diffusion_attention.py

```diff
@@ -0,0 +1,246 @@
+
+        # 1 x H x D -> 1 x H x 1 x D
+        height_emb = height_emb.unsqueeze(2)
+
+        width_emb = self.width_emb(torch.arange(self.width, device=index.device).view(1, self.width))
+
+        # 1 x W x D -> 1 x 1 x W x D
+        width_emb = width_emb.unsqueeze(1)
+
+        pos_emb = height_emb + width_emb
+
+        # 1 x H x W x D -> 1 x L xD
+        pos_emb = pos_emb.view(1, self.height * self.width, -1)
+
+        emb = emb + pos_emb[:, : emb.shape[1], :]
+
+        return emb
+
+
+class VQDiffusionTransformerBlock(nn.Module):
+    """
+    A basic transformer block modified to take the timestep of the diffusion process as additional input.
+    """
+
+    def __init__(
+        self,
+        dim: int,
+        n_heads: int,
+        d_head: int,
+        context_dim: int,
+        diffusion_steps: int,
+        dropout=0.0,
+    ):
+        super().__init__()
+
+        self.attn1 = CrossAttention(
+            query_dim=dim, heads=n_heads, dim_head=d_head, dropout=dropout, bias=True
+        )  # is a self-attention
+        self.ff = VQDiffusionFeedForward(dim=dim, dropout=dropout)
+        self.attn2 = CrossAttention(
+            query_dim=dim, context_dim=context_dim, heads=n_heads, dim_head=d_head, dropout=dropout, bias=True
+        )
+        self.norm1 = AdaLayerNorm(dim, diffusion_steps)
+        self.norm2 = AdaLayerNorm(dim, diffusion_steps)
+        self.norm3 = nn.LayerNorm(dim)
+
+    def forward(self, hidden_states, context, timestep):
+        hidden_states = self.attn1(self.norm1(hidden_states, timestep)) + hidden_states
+        hidden_states = self.attn2(self.norm2(hidden_states, timestep), context=context) + hidden_states
+        hidden_states = self.ff(self.norm3(hidden_states)) + hidden_states
+        return hidden_states
+
+
+class VQDiffusionFeedForward(nn.Module):
+    def __init__(self, dim: int, dim_out: Optional[int] = None, mult: int = 4, dropout: float = 0.0):
+        super().__init__()
+        inner_dim = int(dim * mult)
+        dim_out = dim_out if dim_out is not None else dim
+
+        self.net = nn.Sequential(nn.Linear(dim, inner_dim), GELU2(), nn.Linear(inner_dim, dim), nn.Dropout(dropout))
+
+    def forward(self, hidden_states):
+        return self.net(hidden_states)
+
+
+class AdaLayerNorm(nn.Module):
+    """
+    Norm layer modified to incorporate timestep embeddings.
+    """
+
+    def __init__(self, embedding_dim, num_embeddings):
+        super().__init__()
+        self.emb = nn.Embedding(num_embeddings, embedding_dim)
+        self.silu = nn.SiLU()
+        self.linear = nn.Linear(embedding_dim, embedding_dim * 2)
+        self.norm = nn.LayerNorm(embedding_dim, elementwise_affine=False)
+
+    def forward(self, x, timestep):
+        emb = self.linear(self.silu(self.emb(timestep)))
+        scale, shift = torch.chunk(emb, 2)
+        x = self.norm(x) * (1 + scale) + shift
+        return x
+
+
+class GELU2(nn.Module):
```

Comment:
Could we add this to https://github.com/huggingface/diffusers/blob/main/src/diffusers/models/attention.py ? :-)